### PR TITLE
264 - Added versionInfo endpoint to make the Marklogic Code Version, …

### PR DESCRIPTION
**Pull request by clarkepeterf**
…Date Version, Server Version, Database (Tenant) Name available

### CODE REVIEW

**Comment by clarkepeterf** [comment]
@brent-hartwig mind reviewing this PR?

It makes an endpoint available to get verison info for #264

This does require a small bit of maintenance - there is a new constant with the code version in appConstants.js which requires us to update the version constant each release

File: src/main/ml-modules/root/ds/lux/versionInfo.mjs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+handleRequest(function () {
+  return {
+    codeVersion: CODE_VERSION,
+    dataVersion: _getDataConversionDate(),
+    mlVersion: xdmp.version(),
+    databaseName: xdmp.databaseName(xdmp.database()),
+  };
+});
+
+function _getDataConversionDate() {
```
Comment (brent-hartwig) [inline src/main/ml-modules/root/ds/lux/versionInfo.mjs:13]: Does it matter which document you get?  Here's an alternative that also doesn't throw an exception upon experiencing an issue.

```javascript
function _getDataConversionDate() {
  try {
    return cts
      .doc(fn.subsequence(cts.uris(), 1, 1))
      .xpath('/admin/conversion-date');
  } catch (e) {
    return 'error';
  }
}
```

**Comment by brent-hartwig** [comment]
@clarkepeterf, what would you think about deriving the version from `git`?  In my latest commit:

1. There is still a hard-coded version number, but it specifies the default.  The default is provided by the `defaultCodeVersion` property.
2. `version.sh` is able to return `[tagName]-[commitCount]-[commitHash]`, which presently returns `v1.24.0-1-g61ae787` in my environment.  I'm sure we could get other info from `git` and could update the deployment procedure if there's something for the deployer to do in their environment to get their clone ready to serve up the expected value.
3. The build script sets `codeVersion` to the default value.
4. The `determineCodeVersion` Gradle task is able to override the default using the output from `version.sh`.
5. The `preprocessMarkLogicConfigurationFiles` Gradle task was renamed to `preprocessRuntimeConfigAndModules`.  It sets and resolves the `@@codeVersion@@` token, which `appConstants.mjs` was updated to use.
6. The `mlModulesPaths` property was brought back to tell ML Gradle to deploy the modules from the build directory.

If you like this, let's create a CFT PR for the new props.

**Comment by brent-hartwig** [comment]
@clarkepeterf, please document the endpoint in lux-backend-api-usage.md

**Comment by brent-hartwig** [comment]
Bonus points to add to the Postman export

**Comment by brent-hartwig** [comment]
@clarkepeterf, I just encountered a `XDMP-CONFLICTINGUPDATES` error during deployment.  I'll look at it after lunch.

**Comment by brent-hartwig** [comment]
@clarkepeterf, okay, I made a couple fixes.  Working as expected now.  Below is the return from the new endpoint.  Please take from here (including if you want to stick with your original impl).

```json
{
    "codeVersion": "v1.24.0-6-gcee90d3",
    "dataVersion": "2024-04-17T19:16:17.877022",
    "mlVersion": "11.3.0",
    "databaseName": "lux-content"
}
```

**Comment by clarkepeterf** [comment]
@brent-hartwig I like doing the version from the git tag. I guess we should still try to update the default version in the code before each release. But we should also be using tags, so it should be updated from the tag. And the short ref after the version will help differentiate when we deploy feature commits that don't have a tag.

**Comment by clarkepeterf** [comment]
@brent-hartwig I will update the docs and postman config

**Comment by clarkepeterf** [comment]
@brent-hartwig If you would like to review I made the following changes:

1. Used your suggested `cts.uris()` approach to get a random document for the data conversion date.
2. Removed the --long option from the call to `git describe --tags` - this way if the current revision is where the actaul version tag is, we just see the version tag name. If not, we will see the extra suffix for number of commits and ID
3. Updated the backend api doc
4. Updated postman config

**Comment by brent-hartwig** [comment]
Looks great, @clarkepeterf.  I'm not sure why I ended up with a final commit as it looks like all your edits but so long as I didn't mess anything up... merge this puppy!

**Pull request merged**